R/cluster.R

```diff
@@ -0,0 +1,92 @@
+#' Clustering wrapper
+#'
+#' This function returns a \code{SummarizedExperiment} with clustering 
+#'   information in its colData or rowData
+#'
+#' @param x A
+#'   \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}}
+#'   object.
+#'
+#' @param BLUSPARAM A 
+#'   \code{\link[https://rdrr.io/github/LTLA/bluster/man/BlusterParam-class.html]{BlusterParam}}
+#'   object.
+#'
+#' @param assay.type A single character value for selecting the
+#'   \code{\link[SummarizedExperiment:SummarizedExperiment-class]{assay}}
+#'   to use for identifying dominant taxa.
+#'
+#' @param assay_name a single \code{character} value for specifying which
+#'   assay to use for calculation.
+#'   (Please use \code{assay.type} instead. At some point \code{assay_name}
+#'   will be disabled.)
+#'  
+#'  @param MARGIN A single numeric value for selecting if the clustering is done
+#'   row-wise / for features (1) or column-wise / for samples (2). Must be \code{1} or
+#'   \code{2}. (By default: \code{MARGIN = 1}) 
+#'   
+#' @param full A logical scalar indicating whether the clustering statistics 
+#'   from both steps should be put in metadata
+#'   
+#' @details
+#' When setting \code{full=TRUE}, the clustering information will be stored in
+#' the metadata of the object.
+#'
+#' @author Basil Courbayre
+#'
+#' @examples
+#' data(GlobalPatterns)
+#' x <- GlobalPatterns
+#'
+#' # Cluster on rows using Kmeans
+#' tse <- cluster(tse, KmeansParam(centers = 3))
+#'
+NULL
+
+#' @rdname cluster
+#' @export
+setGeneric("cluster", signature = c("x"),
+           function(x, BLUSPARAM, assay.type = assay_name, assay_name = "counts", 
+                    MARGIN=1, full = FALSE)
+               standardGeneric("cluster"))
+
+#' @rdname cluster
+#' @export
+#' @importFrom bluster clusterRows
+setMethod("cluster", signature = c(x = "SummarizedExperiment"),
+    function(x, BLUSPARAM, assay.type = assay_name, assay_name = "counts", 
+             MARGIN=1, full = FALSE) {
+        # Checking parameters
+        .check_assay_present(assay.type, x)
+        .check_margin(MARGIN)
+        
+        assay <- assay(x, assay.type)
+        
+        # Transpose if clustering on the columns
+        if( MARGIN != 1 ){
+            assay <- t(assay)
+        }
+        
+        result <- clusterRows(assay, BLUSPARAM, full)
+        
+        if (full) {
+            clusters <- result$clusters
+            metadata(x)[["clusters"]] <- result$objects
+        } else {
+            clusters <- result
+        }
+        
+        if (MARGIN == 1) {
+            rowData(x)$clusters <- clusters
+        } else {
+            colData(x)$clusters <- clusters
+        }
+        
+        x
+    }
+)
+
+.check_margin <- function(MARGIN) {
+    if( !is.numeric(MARGIN) && !MARGIN %in% c(1, 2) ){
+        stop("'MARGIN' must be 1 or 2.", call. = FALSE)
```

Comment:
In getExperimentCrossAssociation.R we also use "Samples" and "Features"

In transformCounts we use "samples" and "features"

in addition to MARGIN = 1 or MARGIN = 2.

Should do the same here.

Also harmonize, we should use the same spelling form everywhere. I would go with the lowercase?

Reply:
Utilize utils.R --> we probably have this check already there

Reply:
I haven't seen this in  utils.R but I can add it if you agree.

Reply:
Well, that is true --> add and change also getExperimentCrossAssociation's check accordingly

Reply:
I haven't found any checks for the MARGIN in test-5transformCounts and test-5getExperimentCrossAssociation. 
I also checked test-4splitOn but found nothing.
Should I just add my function to utils.R and check that the value of MARGIN is either "samples", "features", "rows", "cols", 1 or 2?

Reply:
Ahh, sorry. I meant https://github.com/microbiome/mia/blob/0cd430a34d9d7bd4c3ef474e4535c179dd1c650a/R/transformCounts.R#L271

BAsed on this, create a .check_margin function --> it could also do the conversion between columns --> 2 https://github.com/microbiome/mia/blob/0cd430a34d9d7bd4c3ef474e4535c179dd1c650a/R/transformCounts.R#L429

and it return FALSE or correct MARGIN? Then we could modify this allowed MARGIN values from one place